### PR TITLE
Add caching? helper method

### PR DESCRIPTION
**Pull request by joelhawksley**
## Problem

Caching something that shouldn't be cached is a potential source of bugs and security vulnerabilities. For example, one could write a form helper that outputs a request-specific auth token, only for the helper to be used inside of a `cache` block.

## Proposed solution

In the GitHub application, we implemented a caching? method and used it to raise an error if a specific code path is being cached that we don't want to be cached.

I've credited its original author, @btoews.

Co-authored-by: Ben Toews <mastahyeti@gmail.com>
Co-authored-by: John Hawthorn <jhawthorn@github.com>

### CODE REVIEW

**Comment by eileencodes** [comment]
I've been thinking about this PR and I'm not entirely convinced it's necessary or useful to other apps. If the problem is that we don't want to cache sensitive data then we shouldn't add caching blocks for that. Am I missing a piece of how this is used that results in sensitive data getting cached accidentally? The code author would still need to know what data is considered sensitive and need to know to raise if there was sensitive data they didn't want cached.

**Comment by rafaelfranca** [comment]
I think one of the problem is that something people don't know they are inside a cached template/partial but this would still be true even if we add this method. People would still accidentally add sensitive data inside cached blocks. This helper only makes easy to opt-out of the cache fragment and instead of having two different templates for a cached partial and an uncached partial, you only have one. I'm not sure it is worth a fiber-local.

**Comment by joelhawksley** [comment]
@eileencodes @rafaelfranca thanks for taking the time to review. I'll provide a more detailed example to explain why I think this feature is generally useful.

We have a custom form helper (let's call it `button_form`) for building buttons that are forms. Part of this helper includes building CSRF tokens. In order to prevent developers from accidentally using `button_form` inside a `cache` block, we raise an error if we're currently caching:

```ruby
# Build a button that is a form
#
# Returns an HTML String.
def button_form(name, options = {}, html_options = {})
  # Ensure that we aren't caching CSRF tokens.
  raise NoCachingError if caching?

  # build the form, including CSRF tokens
end
```

At least in our application, it can be difficult to reason about the render stack to be certain that a given line of code won't execute inside a `cache` block. I think this kind of helper is a good tool to have for building safety mechanisms like the given example.

**Comment by rafaelfranca** [comment]
We don't use much fragment cache for me to care about this solution, but if others think it is fine it is ok to me. About the implementation, I'd prefer if we add a Runtime Registry in Action View using `ActiveSupport::PerThreadRegistry` instead of using the fiber local directly.

**Comment by kaspth** [comment]
Could also consider adding an `uncachable!` method to auto-raise that `NoCachingError` so users wouldn't have to define their own exception. Think we could use `__callee__` within the exception to extract e.g. `button_form` for the message.

**Comment by joelhawksley** [comment]
@rafaelfranca @kaspth thanks for the reviews. I rewrote this PR to use `ActiveSupport::PerThreadRegistry`. 

@kaspth what do you think about having _both_ `caching?` and `uncachable!`?

**Comment by joelhawksley** [comment]
@kaspth friendly bump 😄

**Comment by kaspth** [comment]
> @kaspth what do you think about having _both_ `caching?` and `uncachable!`?

Sorry I wasn't clear, I meant that we'd have both yeah.

File: actionview/lib/action_view/caching_registry.rb [hunk 1]
```diff
@@ -0,0 +1,10 @@
+# frozen_string_literal: true
+
+module ActionView
+  # = Action View Caching registry
```
Comment (kaspth) [inline actionview/lib/action_view/caching_registry.rb:4]: Do we need this comment when the registry is nodoc'ed?

File: actionview/lib/action_view/helpers/cache_helper.rb [hunk 2]
```diff
@@ -165,15 +167,29 @@ module CacheHelper
...
         end
 
         nil
       end
 
+      # Whether the current code path is being cached.
+      #
+      #   <% raise StandardError.new("Caching private data!") if caching? %>
+      def caching?
+        !!ActionView::CachingRegistry.is_caching
```
Comment (kaspth) [inline actionview/lib/action_view/helpers/cache_helper.rb:190]: We generally just let Ruby's semantics follow through, so no `!!` please.

```suggestion
        ActionView::CachingRegistry.is_caching
```

File: actionview/lib/action_view/helpers/cache_helper.rb [hunk 3]
```diff
@@ -165,15 +167,29 @@ module CacheHelper
...
-          safe_concat(fragment_for(cache_fragment_name(name, **name_options), options, &block))
+          begin
+            caching_was = ActionView::CachingRegistry.is_caching
+            ActionView::CachingRegistry.is_caching = true
+
+            name_options = options.slice(:skip_digest)
+            safe_concat(fragment_for(cache_fragment_name(name, **name_options), options, &block))
+          ensure
+            ActionView::CachingRegistry.is_caching = caching_was
+          end
```
Comment (kaspth) [inline actionview/lib/action_view/helpers/cache_helper.rb:178]: The `CacheHelper` shouldn't track the registry state manually, let's extract a helper:

```suggestion
          ActionView::CachingRegistry.track_caching do
            name_options = options.slice(:skip_digest)
            safe_concat(fragment_for(cache_fragment_name(name, **name_options), options, &block))
          end
```

File: actionview/lib/action_view/helpers/cache_helper.rb [hunk 4]
```diff
@@ -165,15 +167,29 @@ module CacheHelper
...
         else
           yield
         end
 
         nil
       end
 
+      # Whether the current code path is being cached.
+      #
+      #   <% raise StandardError.new("Caching private data!") if caching? %>
```
Comment (kaspth) [inline actionview/lib/action_view/helpers/cache_helper.rb:188]: This is too vague and doesn't use similar language to the rest of the documentation here. Maybe this?

```suggestion
      # Returns whether the current view fragment is within a +cache+ block.
      #
      # Useful when certain fragments aren't cacheable:
      #
      #   <% cache project do %>
      #     <% raise StandardError, "Caching private data!" if caching? %>
      #   <% end %>
```

File: actionview/lib/action_view/caching_registry.rb [hunk 5]
```diff
@@ -0,0 +1,10 @@
...
+  class CachingRegistry #:nodoc:
+    extend ActiveSupport::PerThreadRegistry
+
+    attr_accessor :is_caching
```
Comment (kaspth) [inline actionview/lib/action_view/caching_registry.rb:8]: Let's add the extra line to get the predicate method:

```suggestion
    attr_accessor :caching
    alias caching? caching
```

**Review by kaspth — COMMENTED** [review]
Sweet, getting there!

Let's just define the `CachingRegistry` inline in `CacheHelper` right below the `private` marker. It doesn't warrant being a top-level file at the `action_view/lib/*` root.

File: actionview/CHANGELOG.md [hunk 6]
```diff
@@ -24,6 +24,10 @@
 
     *Dirkjan Bussink*
 
+*   Add `caching?` helper that returns whether the current code path is being cached.
+
+    *Ben Toews*, *John Hawthorn*, *Kasper Timm Hansen*, *Joel Hawksley*
+
```
Comment (kaspth) [inline actionview/CHANGELOG.md:30]: New entries go at the top of the file.

File: actionview/lib/action_view/helpers/cache_helper.rb [hunk 7]
```diff
@@ -165,15 +165,28 @@ module CacheHelper
...
+            safe_concat(fragment_for(cache_fragment_name(name, **name_options), options, &block))
+          end
         else
           yield
         end
 
         nil
       end
 
+      # Returns whether the current view fragment is within a +cache+ block.
```
Comment (kaspth) [inline actionview/lib/action_view/helpers/cache_helper.rb:179]: If we want `uncachable!` for this too, I'd define an exception atop this helper:

```ruby
module CacheHelper
  class UncachableFragmentError < StandardError; end
  …
```

And then define the method above `caching?` like this:

```ruby
# Raises +UncachableFragmentError+ when called from within a +cache+ block.
#
# Useful to denote helper methods that can't participate in fragment caching:
#
#   def project_name_with_time(project)
#     uncachable!
#     "#{project.name} - #{Time.now}"
#   end
#
#   # Which will then raise if used within a +cache+ block:
#   <% cache project do %>
#     <%= project_name_with_time(project) %>
#   <% end %>
def uncachable!
  raise UncachableFragmentError, "can't be fragment cached" if caching?
end
```

**Review by kaspth — APPROVED** [review]
Looks like the changelog entry just needs to be moved and then we could ship this now — we can also tackle the `uncachable!` method here if you'd want.

File: actionview/test/fixtures/test/uncacheable.html.erb [hunk 8]
```diff
@@ -0,0 +1,3 @@
+<%= cache "uhoh" do %>
```
Comment (kaspth) [inline actionview/test/fixtures/test/uncacheable.html.erb:1]: uhoh indeed, love it 😂

**Review by kaspth — APPROVED** [review]
Rad!

**Comment by kaspth** [comment]
I know it was marked as a draft PR, but I couldn't see anything else that was missing, so I went ahead. If that was erroneous, I'll be around so we can tackle it in a follow up PR. Thanks for working on this, I'm really happy with how it turned out! ❤️

**Comment by joelhawksley** [comment]
@kaspth Ha! I was just waiting for CI to pass before moving it out of draft.

Thanks for your help landing this PR, I really appreciate your thoughtful feedback ❤️

**Pull request merged**